From a2835a7de9f5e59f066fbefdf09d8beac15453e7 Mon Sep 17 00:00:00 2001
From: farah kassabri <farah.kassabri@intel.com>
Date: Wed, 27 Nov 2019 21:56:18 -0500
Subject: [PATCH] Dont't try to allocate continuous memory for 4GB or more.

struct scatterlist length field defined as unsigned int and cannot store value of 0x100000000 or more.
In that case, even if we managed to allocate continuous 4GB, we create sg list with one entry,
the length field of that entry will store 0, and when we call gen_pool_free specifying sg->length
it'll free no memory of the heap pool chunk.
---
 drivers/staging/android/ion/ion_chunk_heap.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/staging/android/ion/ion_chunk_heap.c b/drivers/staging/android/ion/ion_chunk_heap.c
index bf6ca7f..d522a0d 100644
--- a/drivers/staging/android/ion/ion_chunk_heap.c
+++ b/drivers/staging/android/ion/ion_chunk_heap.c
@@ -16,6 +16,8 @@
 
 #include "ion.h"
 
+#define MAX_CONT_BUFF_SIZE 0x100000000
+
 struct ion_chunk_heap {
 	struct ion_heap heap;
 	struct gen_pool *pool;
@@ -33,6 +35,9 @@ static int ion_chunk_heap_allocate_contig(struct ion_chunk_heap *chunk_heap,
 	unsigned long paddr;
 	int rc;
 
+	if (allocated_size >= MAX_CONT_BUFF_SIZE)
+		return -EINVAL;
+
 	paddr = gen_pool_alloc_algo(chunk_heap->pool,
 				    allocated_size,
 				    gen_pool_first_fit_align,
-- 
1.8.3.1

